### PR TITLE
Local Reynolds Feature: Re_x per surface node for BL state

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The global Reynolds number (Re) is a single scalar for the entire sample. But boundary layer physics varies locally along the surface — the local Reynolds number Re_x = Re * (x_local / chord) determines whether the boundary layer is laminar, transitional, or turbulent at each point. This Re_x signal is critical for predicting surface pressure distribution accuracy, especially at OOD Reynolds numbers.

**Physics motivation:** The B-GNN paper (arXiv:2503.18638, Zhang et al., March 2025) demonstrated that providing local Re_x alongside inviscid Cp as input features improved OOD boundary layer predictions by 88%. Our inviscid Cp residual target experiment just failed, but providing Re_x as an **input feature** (not target modification) follows the proven pattern of physics-informed features that work (wake deficit -4.1%, TE coord frame -5.4%).

**Key insight:** The model currently receives Re as a global scalar broadcast to all nodes. It must learn to convert this to local BL state for each surface position. By providing Re_x = Re * (chord_fraction), we give the model a physics-grounded local signal that directly encodes where transition/separation is likely. This is especially valuable for:
- **p_re (OOD Reynolds):** The model sees new Re values — but Re_x gives graded local information
- **p_tan:** Wake interference changes effective Re locally on the aft foil

### Step 1: Add argument

```python
parser.add_argument('--local_re_feature', action='store_true',
                    help='Add local Reynolds number Re_x = Re * chord_fraction as surface feature')
```

### Step 2: Compute Re_x per node

For each surface node, compute the chord-normalized distance from leading edge and multiply by the sample's Reynolds number:

```python
def compute_local_re(pos, surf_mask, re_value, leading_edge_pos):
    """Compute local Reynolds number for each node.
    
    Args:
        pos: [N, 2] node positions
        surf_mask: [N] boolean
        re_value: scalar Reynolds number for this sample
        leading_edge_pos: [2] position of leading edge
    Returns:
        local_re: [N, 1] — Re_x for surface nodes, 0 for volume nodes
    """
    # Distance from leading edge along chord direction
    dx = pos[:, 0] - leading_edge_pos[0]  # Streamwise distance
    chord_frac = dx.clamp(min=0) / (dx.max() - dx.min() + 1e-8)  # 0 at LE, ~1 at TE
    
    # Local Re_x = Re * chord_fraction
    re_x = re_value * chord_frac
    
    # Normalize to reasonable range: log(Re_x + 1) / log(Re + 1)
    re_x_norm = torch.log1p(re_x) / (torch.log1p(torch.tensor(re_value)) + 1e-8)
    
    # Zero for non-surface nodes
    local_re = torch.zeros(pos.shape[0], 1, device=pos.device)
    local_re[surf_mask, 0] = re_x_norm[surf_mask]
    
    return local_re
```

**For tandem samples:** Compute separately for fore and aft foils. The aft-foil's effective Re_x is modified by the wake — you can approximate this as `Re_x_aft = Re_x * wake_reduction_factor` where wake_reduction_factor = 0.7-0.9 (rough estimate, or use the wake_deficit_feature if available). Or just compute raw Re_x without wake correction — the model can learn the correction.

### Step 3: Identify Re and leading edge

- **Re value:** Should be in the sample metadata or global features. Check how `re_stratified_sampling` accesses Re — it's likely in the data dict.
- **Leading edge position:** The node with the minimum x-coordinate on the surface, or computed from the NACA geometry features.

### Step 4: Concatenate as input feature

```python
if args.local_re_feature:
    local_re = compute_local_re(pos, surf_mask, re_value, le_pos)
    x = torch.cat([x, local_re], dim=-1)
```

### Step 5: Run 2 seeds

```bash
cd cfd_tandemfoil && python train.py \
  --agent frieren --wandb_name "frieren/local-re-s42" --seed 42 \
  --wandb_group local-reynolds-feature \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling \
  --local_re_feature

# Repeat for seed 73: --seed 73 --wandb_name "frieren/local-re-s73"
```

## Baseline

| Metric | Baseline | Target to beat |
|--------|----------|----------------|
| p_in | 11.742 | < 11.742 |
| p_oodc | 7.643 | < 7.643 |
| p_tan | 27.874 | < 27.874 |
| p_re | 6.419 | < 6.419 |

- **Baseline W&B runs:** k5qwvce4 (seed 42), 7oa5xfhi (seed 73)
- **val/loss baseline:** ~0.37

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2331 because ADVISOR: Closing based on W&B results (z21pbxdo, g7355k1g) — local Re_x feature doesn't provide a clear improvement:

| Metric | 2-seed avg | Baseline | Δ |
|--------|------------|----------|---|
| p_in | 12.366 | 11.742 | +5.3% ❌ |
| p_oodc | 7.592 | 7.643 | -0.7% ≈ |
| p_tan | 28.045 | 27.874 | +0.6% ≈ |
| p_re | 6.591 | 6.419 | +2.7% ❌ |

The p_oodc improvement is marginal (-0.7%) and seed-unstable (7.32 vs 7.87). Meanwhile p_in regresses +5.3% — same pattern as the panel Cp feature.

**Root cause:** The local Re_x = Re × chord_fraction is essentially a scaled version of the chord-position feature, which we already know is redundant with TE coordinate frame. The Re value is already a global feature — multiplying it by position doesn't add genuinely new information that the model can't compute internally.

**Lesson:** Local Re is too simple an encoding. The model already has global Re and local position separately — the product doesn't help because transformers can learn multiplicative interactions.

Moving you to a fresh experiment..

**Comment by tcapelle** [comment]
STUDENT frieren:

## Results

Both seeds completed at 144–145 epochs.

| Metric | Seed 42 | Seed 73 | 2-seed avg | Baseline | Δ avg |
|--------|---------|---------|------------|----------|-------|
| val/loss | 0.3824 | 0.3796 | 0.3810 | ~0.37 | — |
| p_in | 12.32 | 12.42 | 12.37 | 11.742 | +5.3% ❌ |
| p_tan | 28.6 | 27.5 | 28.05 | 27.874 | +0.6% ≈ |
| p_oodc | 7.32 | 7.87 | 7.60 | 7.643 | -0.6% ≈ |
| p_re | 6.69 | 6.50 | 6.60 | 6.419 | +2.8% ❌ |

**W&B runs:** z21pbxdo (seed 42), g7355k1g (seed 73), group: local-reynolds-feature

**Command used (seed 42):**
```bash
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent frieren --wandb_name "frieren/local-re-s42" --seed 42 \
  --wandb_group local-reynolds-feature \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling \
  --local_re_feature
```

**Peak memory:** ~78 GB (GPU 0), ~80 GB (GPU 1)

### What happened

The local Re_x feature did not improve the baseline. p_in regressed +5.3% and p_re regressed +2.8%. p_oodc showed a marginal -0.6% average but this is seed-unstable (7.32 vs 7.87). p_tan is essentially unchanged.

**Root cause analysis:**
- **Redundancy with existing features:** The model already sees `log_re` (global scalar, index 13) and `pos` (x,y coordinates). Re_x = Re × chord_fraction is the product of these two, and transformers can learn multiplicative interactions implicitly — the explicit encoding doesn't add genuinely new information.
- **Same failure pattern as TE coordinate frame redundancy:** Adding Re_x mirrors a scaled version of the x-position feature we already provide, similar to how foil2_dist was redundant once TE coord frame was added.
- **Volume nodes get zero:** The 0/non-zero discontinuity between surface and volume nodes may add a spurious signal at the boundary rather than helping.
- **p_re wasn't the main OOD failure mode:** The val_ood_re split tests generalizing to new Re values, and the model already handles this well (p_re=6.42 baseline). Adding Re_x didn't help because the model's main bottleneck for p_re is something else.

### Suggested follow-ups

1. **Wall-normal distance feature:** Unlike Re_x, wall-normal distance (minimum DSDF over surface nodes) provides information about boundary layer thickness that isn't trivially computable from existing features.
2. **Learnable physics embeddings:** Instead of hand-crafting Re_x, use a small MLP to generate physics embeddings from (Re, AoA, NACA) that condition node features — gives the model latitude to learn its own useful Re encodings.
3. **Foil-boundary interaction features:** Encode which surface nodes are near the wake of the fore foil (using wake_deficit distance) — this is a non-trivial geometric interaction not captured by existing features.
